### PR TITLE
Implemented #194.

### PR DESCRIPTION
**Pull request by redboltz**
Send multiple messages in `queue_` all at once.
Size limit is not implemented yet.

### CODE REVIEW

**Comment by redboltz** [comment]
Performance checking now...

File: include/mqtt/endpoint.hpp [hunk 1]
```diff
@@ -5946,17 +5946,30 @@ class endpoint : public std::enable_shared_from_this<endpoint<Socket, Mutex, Loc
     }
 
     void do_async_write() {
-        auto const& elem = queue_.front();
-        auto const& mv = elem.message();
+        std::vector<as::const_buffer> buf;
+        std::vector<async_handler_t> handlers;
```
Comment (jonesmz) [inline include/mqtt/endpoint.hpp]: Don't forget std::vector<>::reserve()

**Comment by redboltz** [comment]
Removing from queue_ requires to change.

File: include/mqtt/endpoint.hpp (hunk 1 above)
Comment (redboltz) [reply include/mqtt/endpoint.hpp]: I think that there is no easy way to know the number of `as::const_buffer`.  See https://github.com/redboltz/mqtt_cpp/blob/master/include/mqtt/message.hpp#L741

In order to get the number of chunks, need to add the new member function such as `num_of_const_buffers()` to each message type. And call it via visitor mechanism.

Comment (redboltz) [reply include/mqtt/endpoint.hpp]: I will add by this approach.

Comment (redboltz) [reply include/mqtt/endpoint.hpp]: I added.
This approach requires one more loop for calculating the number of const buffer sequence. Call getting the number of const buffer sequence function for each element via visitor.

I'm not sure that this approach is really more efficient than allow vector expansion (reallocation).

Comment (redboltz) [reply include/mqtt/endpoint.hpp]: I believe that calling reserve approach is efficient even if additional loop and visitor call is executed. vector reallocation could invoke system calls. It is high cost.

**Comment by redboltz** [comment]
The default value of `max_queue_send_count_` is 1. That means the same as the original behavior. `async_write()` calls one by one messages.

User can set the count using the following API.

```cpp
     /**
     * @brief Set maximum number of queued message sending.
     *        When async message sending function called during asynchronous
     *        processing, the message is enqueued. When current asynchronous
     *        message is processed, then concatenate queued messages and
     *        send it.
     *        This value limits the maximum number of concatenating messages.
     *        The default value is 1.
     *
     * @param count maximum number of queued message sending. 0 means infinity.
     *
     */
    void set_max_queue_send_count(std::size_t count) {
        max_queue_send_count_ = count;
    }
```

When I set the count to 0 (infinity), then I got some performance problem. I need to analyze it before merging. I can't user the performance testing environment this week. I will restart the checking next week.

**Comment by redboltz** [comment]
I've checked the performance. When I set the count to 1, then there is no observable difference from the original code. That means we can ignore the additional code cost by the PR.

When I set the count to 512, and tested with very high load e.g. 50000 publish/sec, I observed the behavior change, but no significant throughput improvement is observed.
I think that it depends on environment.

Anyway, if we use with the default setting, there is no difference from the original behavior. If users can tuning the behavior with setting the count and/or size.

I merge the PR.

**Pull request merged**